engines/buried/sound.cpp

```diff
@@ -0,0 +1,761 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "audio/audiostream.h"
+#include "audio/mixer.h"
+#include "audio/decoders/wave.h"
+#include "common/archive.h"
+#include "common/events.h"
+#include "common/system.h"
+
+#include "buried/buried.h"
+#include "buried/graphics.h"
+#include "buried/resources.h"
+#include "buried/sound.h"
+
+namespace Buried {
+
+#define TIMED_EFFECT_NONE   0x00
+#define TIMED_EFFECT_VOLUME 0x01
+
+#define SOUND_FLAG_DESTROY_AFTER_COMPLETION 0x01
+
+static inline int clipVolume(int volume) {
+	return CLIP<int>(volume, 0, Audio::Mixer::kMaxChannelVolume);
+}
+
+SoundManager::SoundManager(BuriedEngine *vm) : _vm(vm) {
+	_fileIDFootsteps = -1;
+	_lastAmbient = 1;
+	startup();
+}
+
+SoundManager::~SoundManager() {
+	for (int i = 0; i < kMaxSounds; i++)
+		delete _soundData[i];
+}
+
+bool SoundManager::startup() {
+	_paused = false;
+
+	for (int i = 0; i < kMaxSounds; i++)
+		_soundData[i] = new Sound();
+
+	return true;
+}
+
+void SoundManager::shutDown() {
+	if (_paused)
+		return;
+
+	for (int i = 0; i < kMaxSounds; i++) {
+		delete _soundData[i];
+		_soundData[i] = new Sound();
+	}
+}
+
+void SoundManager::pause(bool shouldPause) {
+	for (int i = 0; i < kMaxSounds; i++)
+		_soundData[i]->pause(shouldPause);
+}
+
+bool SoundManager::setAmbientSound(const Common::String &fileName, bool fade, byte finalVolumeLevel) {
+	// Determine which of the two ambient tracks to use
+	int newAmbientTrack = (_lastAmbient == 0) ? 1 : 0;
+
+	// If this ambient track is currently in use, stop and kill it now
+	if (_soundData[kAmbientIndexBase + newAmbientTrack]->isPlaying()) {
+		delete _soundData[kAmbientIndexBase + newAmbientTrack];
+		_soundData[kAmbientIndexBase + newAmbientTrack] = new Sound();
+	}
+
+	bool retVal = true;
+
+	if (fileName.empty()) {
+		if (fade) {
+			// Set parameters for the current ambient, if there is one
+			if (_soundData[kAmbientIndexBase + _lastAmbient]->isPlaying()) {
+				_soundData[kAmbientIndexBase + _lastAmbient]->_loop = true;
+				_soundData[kAmbientIndexBase + _lastAmbient]->_timedEffectIndex = TIMED_EFFECT_VOLUME;
+				_soundData[kAmbientIndexBase + _lastAmbient]->_flags = SOUND_FLAG_DESTROY_AFTER_COMPLETION;
+				_soundData[kAmbientIndexBase + _lastAmbient]->_timedEffectSteps = 16;
+				_soundData[kAmbientIndexBase + _lastAmbient]->_timedEffectDelta = -(_soundData[kAmbientIndexBase + _lastAmbient]->_volume / 16);
+				_soundData[kAmbientIndexBase + _lastAmbient]->_timedEffectStart = g_system->getMillis();
+				_soundData[kAmbientIndexBase + _lastAmbient]->_timedEffectRemaining = 2000;
+
+				// Reset parameters for current ambient
+				g_system->getMixer()->setChannelVolume(*_soundData[kAmbientIndexBase + _lastAmbient]->_handle, clipVolume(_soundData[kAmbientIndexBase + _lastAmbient]->_volume << 1));
+				// clone2727: The original resets the loop count, but I don't think that's necessary
```

Comment:
Nothing to do with the current arguments, but it's generally not practice to be putting names of specific developers in coments. Those are better kept to commit messages. These would be better reformatted without the clone2727. Or, pt a more standard TODO/WORKAOUND prefix so we can find them more easily later on.